### PR TITLE
Fix complex run summation in MultiFileNameParser

### LINKED ISSUE #21757: MultipleFileProperty does not work with `1-2+3-4` like string.
`MutlipleFileProperty` does not support complex summation formulas such as `1-2+4` or `1+3-6`.

### PR DESCRIPTION
**Pull request by soininen**
This PR adds the ability to specify complex sum ranges, like `1-3+5` and `2+5-23` to `MultiFileNameParser`.

**To test:**

Try, for example, the `Load` or `LoadAndMerge` algorithms with complex summation both from MantidPlot (or Workbench, if you dare) and from Python.

Fixes #21757.

**Release Notes** 

The fix is mentioned in the Framework release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/release/v3.12.0/framework.rst [hunk 1]
```diff
@@ -65,6 +65,7 @@ Core Functionality
 - Added new classes ``ConfigObserver`` for listening for changes to any configuration property and ``ConfigPropertyObserver`` for listening to changes to an individual config property of interest.
 - Fixed the calculation of scattering length and scattering length squared for :py:obj:`Material <mantid.kernel.Material>`
 - Fixed the behaviour of ``UpdateInstrumentDefinitions.OnStartup`` in the :ref:`properties file <Properties File>`. It was not being used correctly for using the updated ``Facilities.xml`` file.
+- ``MultiFileProperty`` now accepts complex summation ranges for run numbers, such as ``111-113+115`` and ``111-115+123-132``.
```
Comment (SimonHeybrock) [inline docs/source/release/v3.12.0/framework.rst]: Question:
Is it potentially confusing to have `-` for defining a range but `+` combining ranges? Is this convention following any existing behavior?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Looks good overall, sensible tests were added.

See one question below.

Furthermore: Should this new option also be documented somewhere (tooltips, for example?).

File: docs/source/release/v3.12.0/framework.rst (hunk 1 above)
Comment (soininen) [reply docs/source/release/v3.12.0/framework.rst]: `+` and `-` deal with adding (merging) of data. The documentation of `MultipleFileProperty` states that 'Added Ranges and Stepped Added Ranges may be plussed however.' At least to me the behavior of adding 'added ranges' with `+` feels logical, and the use pattern will be common at here at ILL where users frequently merge several files into one workspace.

I'll see if I can improve the documentation. At least `MultipleFileProperty` could include examples of this kind, maybe `Load`, `LoadAndMerge` and some tooltip could be improved as well

Comment (SimonHeybrock) [reply docs/source/release/v3.12.0/framework.rst]: My point is: Is there a risk that having `-` for the range is interpreting as *subtracting* that run from the others? If so, would `:` work better?

Comment (soininen) [reply docs/source/release/v3.12.0/framework.rst]: There is indeed that risk, but `-` is and has always been an 'Added Range', also according to `MultipleFileProperty` documentation. `:` is a range of individual files so it cannot be used for summing. I don't think `n:m+k:l` makes sense, and is explicitly prohibited in `MultipleFileProperty` documentation.

Comment (SimonHeybrock) [reply docs/source/release/v3.12.0/framework.rst]: Good point, I did not know that. I guess then we do not have a choice and just keep it as it is. :)

**Review by SimonHeybrock — APPROVED** [review]

**Comment by martyngigg** [comment]
Test failure is unrelated.

**Pull request merged**